diagrams/sequence-diagrams/file-upload.md

```diff
@@ -1,7 +1,62 @@
-## Uploading a file when completing a form
-
 > [!NOTE]
-> File upload has not yet been implemented, these are ideas / proposals:
+> File upload is under active development, the following is likely to change:
+
+## File Upload Architecture
+
+```mermaid
+---
+title: GOV.UK Forms file upload
+---
+
+graph LR
+
+    user((user<br />filling in<br />form))
+    file[file to<br />upload]
+    browser[web browser]
+
+    user --> browser --http POST--> cloudfront --http POST--> forms --PutObject--> s3 --GetObject--> forms --SendEmail--> ses --send email with attachment(s)----> inbox
+    file --> browser
+
+    subgraph forms_aws [GOV.UK Forms AWS Account]
+        cloudfront[Amazon CloudFront]
+        waf["AWS WAF<br />(Web Application Firewall)"]
+        forms[forms-runner]
+        rds[("Amazon RDS<br/>(Relational<br/>Database<br/>Service)")]
+        s3[\Amazon S3<br />bucket/]
+        guardduty[Amazon GuardDuty<br />S3 Malware Protection]
+        ses["Amazon SES<br />(Simple Email Service)"]
+        sns["Amazon SNS<br />(Simple Notification Service)"]
+        cloudwatch[Amazon<br />CloudWatch]
+
+        cloudfront --inspect request--> waf
+
+        s3 --new object event--> guardduty --tag object--> s3
+        
+        forms --subscribe to topic--> sns --receive delivery notification--> forms
+
+        forms --store completed form<br/>update delivery status--> rds --get completed form--> forms
```

Comment:
Is it worth indicating that we'll delete completed forms too once our retention period has elapsed, or is the too much detail?

Reply:
Yes, I'll mention it here, thanks.